### PR TITLE
Challenge/string_length

### LINKED ISSUE #41: Return the length of a string
Create a program that returns the length of a string

Helpful link- https://www.w3schools.com/python/python_examples.asp

### PR DESCRIPTION
**Pull request by Safiya-hash**
---
name: solution review 
about: A template PR for code review with a checklist
---
Resolves #41 

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/string_length.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+"""
+
+
+def string_length(string: str) -> int:
+    """
+    The function calculates the length of a given string.
+
+    Parameters:
+        Text (str): The string whose length is to be calculated.
```
Comment (ndubuisia) [inline solutions/string_length.py]: I think the variable name(string) when defining the function should match that in the docstring parameters(Text)

File: solutions/string_length.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        >>> string_length("#(!)@") # special characters are counted as strings
+        5
+        >>> string_length("Documenting,testing and debugging")
+        31
+        >>> string_length("    ") # spaces are counted as strings
+        4
+    """
+
+    if not isinstance(string, str):
+        raise TypeError("The input must be a string.")
```
Comment (ndubuisia) [inline solutions/string_length.py]: In the function, TypeError is raised for wrong value input, but AssertionError was documented. Some doctests also failed

**Review by ndubuisia — COMMENTED** [review]
A nice code you've got here @Safiya-hash. Just a few things to consider

File: solutions/tests/test_string_length.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def test_word(self):
+    """It should return the correct length for a word."""
+    self.assertEqual(string_length("Hey"), 3)
+
+
+def test_sentence(self):
+    """It should return the correct length for a sentence with spaces."""
+    self.assertEqual(string_length("Second challenge!"), 18)
```
Comment (ndubuisia) [inline solutions/tests/test_string_length.py]: The actual return value of this test and the expected value are different.

File: solutions/tests/test_string_length.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def test_spaces_as_string(self):
+    """It should return the correct length of spaces in a string."""
+    self.assertEqual(string_length("          "), 10)
+
+
+def test_mixed_characters(self):
+    """It should return the correct length for a string with mixed characters."""
+    self.assertEqual(string_length("Documenting,testing and debugging"), 31)
```
Comment (ndubuisia) [inline solutions/tests/test_string_length.py]: This test failed.

**Review by ndubuisia — DISMISSED** [review]
@Safiya-hash the test is great covering every test category. You just need to correct some failed test and you are good to go.

File: solutions/string_length.py (hunk 1 above)
Comment (Safiya-hash) [reply solutions/string_length.py]: You are definitely right; I was confused to add string in text (text: str) but then changed my mind though forgot to update it. Thanks, I have updated it now

File: solutions/string_length.py (hunk 2 above)
Comment (Safiya-hash) [reply solutions/string_length.py]: Yeah, there was a problem with my counting, I have resolved it now.

File: solutions/tests/test_string_length.py (hunk 3 above)
Comment (Safiya-hash) [reply solutions/tests/test_string_length.py]: Again, the counting problem. I am happy we have this review stage and seeing my work from my groupmates eyes is a blessing.

File: solutions/tests/test_string_length.py (hunk 4 above)
Comment (Safiya-hash) [reply solutions/tests/test_string_length.py]: There was a counting problem here too, resolved.

File: solutions/tests/test_string_length.py [hunk 5]
```diff
@@ -26,8 +26,10 @@ class TestStringLength(unittest.TestCase):
 
 
 def test_empty_string(self):
-    """It should return 0 for an empty string."""
-    self.assertEqual(string_length(""), 0)
+    """It should return nothing for an empty string."""
+    self.assertEqual(
+        string_length(""),
+    )
```
Comment (ndubuisia) [inline solutions/tests/test_string_length.py]: Just wanted to ask why you changed the expected value of this test.

**Review by ndubuisia — COMMENTED** [review]
Great work you have got here @Safiya-hash. I want to clarify a tiny stuff.

**Comment by Safiya-hash** [comment]
My doctest failed for it should return nothing instead it returns zero i
have yet to push my final commit, could you wait until tomorrow for the
final review?

**Comment by ndubuisia** [comment]
For an empty string, the actual value should be 0.
Till tomorrow then

**Comment by Safiya-hash** [comment]
Yeah, i tried it it failed, though i will figure it out. Thank you so much!

**Comment by Safiya-hash** [comment]
@ndubuisia, thanks for your time. I have resolved the issues; you may review it now.

**Comment by ndubuisia** [comment]
Everything is ready to go. Congrats!

**Review by ndubuisia — DISMISSED** [review]
Everything is ready to go. Congrats.

**Review by ndubuisia — DISMISSED** [review]
@Safiya-hash can you please update your branch so that I can proceed with the review.

**Comment by Safiya-hash** [comment]
Can you try again please? @ndubuisia

**Review by ndubuisia — DISMISSED** [review]
Your work is ready to be merged.

**Comment by ndubuisia** [comment]
@Safiya-hash did you update your branch because it is saying that "the merge-base changed after approval".

**Review by ndubuisia — DISMISSED** [review]
Great work

**Review by ndubuisia — DISMISSED** [review]
The function and test passed the checklist and is ready to be merged.

**Review by ndubuisia — DISMISSED** [review]
Congrats your file is ready to be merged.

**Review by ndubuisia — DISMISSED** [review]
A great code you have got here
